imi/src/main/java/org/motechproject/nms/imi/service/impl/CdrFileServiceImpl.java

```diff
@@ -42,167 +59,426 @@
-        File userDirectory = new File(System.getProperty("user.home"));
-        File cdrDirectory = new File(userDirectory, settingsFacade.getProperty(CDR_FILE_DIRECTORY));
-        File file = new File(cdrDirectory, fileInfo.getCdrFile());
+    private void sendProcessDetailFileEvent(FileInfo fileInfo) {
+        Map<String, Object> eventParams = new HashMap<>();
+        eventParams.put(FILE_INFO_PARAM_KEY, fileInfo);
+        MotechEvent motechEvent = new MotechEvent(PROCESS_DETAIL_FILE_SUBJECT, eventParams);
+        eventRelay.sendEventMessage(motechEvent);
+    }
 
-        return parseDetailRecords(file, fileInfo);
+
+    private File localCdrDir() {
+        return new File(settingsFacade.getProperty(LOCAL_CDR_DIR));
     }
 
 
-    private void sendProcessCdrEvent(CallSummaryRecordDto record) {
-        Map<String, Object> eventParams = new HashMap<>();
-        eventParams.put(CSR_PARAM_KEY, record);
-        MotechEvent motechEvent = new MotechEvent(PROCESS_SUMMARY_RECORD, eventParams);
-        eventRelay.sendEventMessage(motechEvent);
+    // Phase 1: verify the file exists, the csv is valid and its record count and checksum match the provided info
+    //          while collecting a list of errors on the go.
+    //          Does not proceed to phase 2 if any error occurred and returns an error
+    @Override
+    public List<String> verifyDetailFileChecksumAndCount(FileInfo fileInfo) {
+        File file = new File(localCdrDir(), fileInfo.getCdrFile());
+
+        List<String> errors = iterateDetailFile(file, fileInfo, Action.PASS1);
+        if (errors.size() > 0) {
+            // Ends up in a HttpStatus.BAD_REQUEST response to IMI which should generate an IMI-OPS alert
+            reportAuditAndThrow(fileInfo.getCdrFile(), errors);
+        }
+
+        // Send a MOTECH event to continue to phase 2 (without timing out the POST from IMI)
+        sendProcessDetailFileEvent(fileInfo);
+
+        return errors;
     }
 
 
+    // Phase 2: Aggregate detail records into summary records
+    //          Validate all CallSummaryRecord fields exist in the database
+    //          Stop processing if errors occurred
     @Override
-    public void processDetailFile(FileInfo fileInfo) {
-        ParseResults parseResults = parseDetailFile(fileInfo);
+    @MotechListener(subjects = { PROCESS_DETAIL_FILE_SUBJECT })
+    public List<String> processDetailFile(MotechEvent event) {
+        FileInfo fileInfo = (FileInfo) event.getParameters().get(FILE_INFO_PARAM_KEY);
+        File file = new File(localCdrDir(), fileInfo.getCdrFile());
+        String fileName = file.getName();
+
+        // Copy the file from the IMI network share (imi.remote_cdr_dir) into local cdr dir (imi.local_cdr_dir)
+        // Another MOTECH node than the one which picked up the HTTP cdrFileNotification might be picking up this
+        // event, so make sure to copy the file locally, if it's not there already.
+
+        if (file.exists()) {
```

Comment:
nvm, i see that the controller calls the scp before pass 1. Would it make sense to have the scp helper as a separate method that gets invoked at different places?